---
title: "HomeAssignment_DataWrangling"
author: "Arpan"
date: "2025-03-19"
output: 
  html_document:
  md_document:
    variant: gfm
---
# Data Wrangling and Tidyverse

Data wrangling is the idea of manipulating and handling the large amount of messy data.
One way to do it is via tidyverse
The tidyverse is a bunch of packages and functions written by the same folks that manage Rstudio. The tidyverse builds upon base R to allow for easier use of large datasets.


```{r Instaling Packages  }
#Installing and loading tidyverse
#install.packages("tidyverse")
library(tidyverse)
```

Tidyverse includes 8 different packages.
Some of the most useful functionality of some tidyverse functions are the following: 
  - Data wrangling & manipulation
    - mutate()
    - select()
    - filter()
    - the pipe %>% 
    - summarise() 
    - group_by()
    - joining 
    - pivotting 
    - Integration with plotting
    
```{r Loading data}
microbiome.fungi<-read.csv("Basics_to_R/Bull_richness.csv")
str(microbiome.fungi)
```
# **Working with Tidyverse**

## Selecting columns

```{r Selecting columns}
#To select certain number of columns that you want to use for data aalysis
microbiome.fungi2 <- select(microbiome.fungi, SampleID, Crop, Compartment:Fungicide, richness)
#Use of colon is to select Compartment through Fungicide, meaning everything in order in between them.
```

## Filtering 
Does the same thing as the subset function does

```{r Filtering}
head(filter(microbiome.fungi2, Treatment == "Conv."))

# A more complex using &
head(filter(microbiome.fungi2, Treatment == "Conv." & Fungicide == "C"))

# Another more complex example using or |
head(filter(microbiome.fungi2, Sample == "A" | Sample == "B")) # samples A or B
```

## Creating new columns
 Mutate function allows us to quickly create new columns.

```{r Mutating columns}
#Creating new column using base R
microbiome.fungi2$logRich <- log(microbiome.fungi2$richness) 

# Create a new column called logRich using mutate function
head(mutate(microbiome.fungi2, logRich = log(richness)))
#logRich is the column name and whatever after = is what we want in that column

# Creating a new column which combines Crop and Treatment 
head(mutate(microbiome.fungi2, Crop_Treatment = paste(Crop, Treatment)))
```
## Using Pipe function
The pipe `%>%`
It combines output from one function to input for another function. We can also combine multiple functions together. It helps to try to visualize what you are doing to your data at each step. Basically the pipe allows you to input the previous data from the right side of the pipe into the function on the left (or next line).

*Lets do everything we have done above using the pipe function*
```{r Using pipe function}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns
  filter(Treatment == "Conv.") %>% # subsetting to only include the conventional treatment
  mutate(logRich = log(richness)) %>% # creating a new column of the log richness
  head() # displaying the first six rows 
```

## Using summarise function
We can use the `summarise()` function to find things like means and standard deviations/errors.
```{r Using summarise}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns
  filter(Treatment == "Conv.") %>% # subsetting to only include the conventional treatment
  mutate(logRich = log(richness)) %>% # creating a new column of the log richness
  summarise(Mean.rich = mean(logRich)) # calculating overall mean log richness within the conventionally managed treatment
```

We can also connect multiple summary statistics here.

```{r Multiple functions}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns
  filter(Treatment == "Conv.") %>% # subsetting to only include the conventional treatment
  mutate(logRich = log(richness)) %>% # creating a new column of the log richness
  summarise(Mean.rich = mean(logRich), # calculating the mean richness, stdeviation, and standard error
            n = n(), # this calculates the number of rows
            sd.dev = sd(logRich)) %>% #calculates SD
  mutate(std.err = sd.dev/sqrt(n)) #Adds new column with standard error
```

To calculate summary statistics by group,we can use the `group_by()` function. 
```{r grouping by variable of interest}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  mutate(logRich = log(richness)) %>% # creating a new column of the log richness
  summarise(Mean.rich = mean(logRich), # calculating the mean richness, stdeviation, and standard error
            n = n(), 
            sd.dev = sd(logRich)) %>%
  mutate(std.err = sd.dev/sqrt(n))
```
## Connecting to plotting
Direct input into ggplot is also possible. In this case we would not have to enter the data statement,we just need to start inputing your aesthetics.
```{r plotting with pipe}
plot1 <- microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  mutate(logRich = log(richness)) %>% # creating a new column of the log richness
  summarise(Mean.rich = mean(logRich), # calculating the mean richness, stdeviation, and standard error
            n = n(), 
            sd.dev = sd(logRich)) %>%
  mutate(std.err = sd.dev/sqrt(n)) %>%
  ggplot(aes(x = Fungicide, y = Mean.rich)) + # adding in a ggplot but the data set need not be specified as pipe does it for us
  geom_bar(stat="identity",aes(fill=Fungicide)) +
  geom_errorbar( aes(x=Fungicide, ymin=Mean.rich-std.err, ymax=Mean.rich+std.err), width=0.4) +
  theme_minimal() +
  xlab("") +
  ylab("Log Richness") +
  facet_wrap(~Treatment)
plot1
```

## Joining 

To match up data by rows based on some criteria, we can easily do this with the joining functions. They include:
  -left_join() 
    - Keep all rows of X and add matching rows from Y. Any rows in Y that donâ€™t match X are excluded.
  -right_join() 
    - reverse of left_join()
  -inner_join() 
    - only keep rows that are common to both X AND Y, remove everything else. 
  -full_join()
    - Keep any columns that are in either X or Y
    
To try these out lets split our data into separate sets including taxonomy and metadata.

Here we are also sampling 100 random rows of our dataset using the `sample_n()` function
```{r}
# selecting just the richness and sample ID
richness <- microbiome.fungi %>%
  select(SampleID, richness)

# selecting columns that don't include the richness
metadata <- microbiome.fungi %>% 
  select(SampleID, Fungicide, Crop, Compartment, GrowthStage, Treatment, Rep, Sample)

head(metadata)
head(richness)


head(left_join(metadata, richness, by = "SampleID")) # adding the richness data to the metadata based on on the common column of sampleID
#if we swap it aroung i.e richness,metadata it does the reverse)
```
## Pivoting

Pivoting is also useful for converting tablefrom wide to long format and back again. We can do this with `pivot_longer()` and `pivot_wider()`

Here is a  demonstratration using `pivot_wider()` assuming that most data is already formatted as long data. Lets explore turning our data set into a wide format, since its in long format.
```{r}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% 
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  summarise(Mean = mean(richness)) # calculates the mean per Treatment and Fungicide 
```
Now I want to calculate the difference between means. The prblem is its hard, we need to subset the data and then make columns for mean of each. for this we can use the pivot_wider()
Wide format - sets the values within the fungicide column into column names
```{r}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns  
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  summarise(Mean = mean(richness)) %>% # calculates the mean per Treatment and Fungicide 
  pivot_wider(names_from = Fungicide, values_from = Mean) # pivot to wide format, take the fungicide names and put in new columns and then take mean values 
```

Easily can take the difference bewteen the fungicide and control now. 
```{r}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns  filter(Class == "Sordariomycetes") %>%
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  summarise(Mean = mean(richness)) %>% # calculates the mean per Treatment and Fungicide 
  pivot_wider(names_from = Fungicide, values_from = Mean) %>% # pivot to wide format
  mutate(diff.fungicide = C - F) # calculate the difference between the means. 
```

Now we can easily calculate the difference between fungicide and control and plot it. 
```{r}
microbiome.fungi %>%
  select(SampleID, Crop, Compartment:Fungicide, richness) %>% # selecting columns  filter(Class == "Sordariomycetes") %>%
  group_by(Treatment, Fungicide) %>% # grouping by treatment and fungicide to later calculate summary stats by group
  summarise(Mean = mean(richness)) %>% # calculates the mean per Treatment and Fungicide 
  pivot_wider(names_from = Fungicide, values_from = Mean) %>% # pivot to wide format
  mutate(diff.fungicide = C - F) %>%  # calculate the difference between the means. 
  ggplot(aes(x = Treatment, y = diff.fungicide)) + # Plot it 
  geom_col() +
  theme_minimal() +
  xlab("") +
  ylab("Difference in average species richness")
```